[processor/tailsampling] Support extensions

csmarchbanks opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
Allow the Tail Sampling Processor to use extensions that implement an interface simply containing NewEvaluator. To use an extension all that needs to be done is specify the extension's type/name such as my_extension or my_extension/my_name in the type field of the policy configuration. When configuration is loaded or updated NewEvaluator will be called.

Extensions can be configured in two places. First, in the extensions section like all other extensions. This is useful for configuration such as if an extension needs to connect to an external service. In addition, configuration can be defined in the policy config under a key matching the type of the extension for anything related to one instance, following the pattern for all other policy types. This is useful when reusing the same extension e.g. different params for different services. For example:
```yaml
type: custom_policy
custom_policy:
  foo: bar
```

What this ends up with is that extensions look exactly like all other policy types with no need for users to know about some sort of extension container.

I have tested this out with an extension in a custom build and it is working well, I am sure we will iterate on the interfaces a bit as we write more. Specifically a way to stop the extension when policies are updated may be necessary to clean up any background processes. @yvrhdn is working on an extension we plan to add to contrib as well.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Part of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31582

csmarchbanks committed:
[processor/tailsampling] Support extensions

Allow the Tail Sampling Processor to use extensions that implement an
interface simply containing NewEvaluator. To use an extension all that
needs to be done is specify the extension's type/name such as
my_extension or my_extension/my_name in the type field of the policy
configuration. When configuration is loaded/updated the extension will
be called in order to create a new evaluator.

Extensions can be configured in two places. First, in the extensions
section like all other extensions. This is useful for configuration such
as if an extension needs to connect to an external service. In addition,
an object can defined under the "extension" key the policy config for
anything related to one instance. This is useful when reusing the same
extension e.g. different params for different services.

yvrhdn commented on processor/tailsamplingprocessor/processor.go:
```diff
@@ -216,22 +216,31 @@ func withRecordPolicy() Option {
 
+	extension, ok := policyExtensions[string(cfg.Type)]
+	if ok {
+		evaluator, err := extension.NewEvaluator(cfg.ExtensionCfg)
+		if err != nil {
+			return nil, fmt.Errorf("unable to load extension %s: %w", string(cfg.Type), err)
+		}
```
What do you think of passing in `Name` and `Type` as well?
- Type could be useful if an extension has multiple policies and wants to switch between them internally.
- Name might be useful in case the extension wants to emit its own metrics/logs

csmarchbanks replied:
I updated the `Extension` interface to now take a `policyName`.

csmarchbanks committed:
Match extension config name to the type

This matches all of the other policy types where the key for the
configuration matches the type of the policy. E.g. you configure
ottl_condition by setting ottl_condition: {...}

csmarchbanks committed:
Pass policy name to NewEvaluator

It can be useful for an instance of an evaluator to know the name of the
policy for debugging or reporting purposes. Since a single type can be
used multiple times with different configuration.

csmarchbanks committed:
Fix lint errors

csmarchbanks committed:
Add changelog

atoulme commented:
@portertech please review as codeowner, thanks!

csmarchbanks commented:
This has codeowner approval now, anything else needed before merge?
